Improve standard dialog examples and documentation (#1053)
enthought/pyface: docs/source/standard_dialogs.rst

.. _standard_dialogs:

================
Standard Dialogs
================

In addition to being able to create custom dialogs based on the |IDialog|
interface, Pyface exposes a number of standard toolkit and OS dialogs.

About Dialog
============

.. figure:: images/about_dialog.png
   :scale: 50
   :alt: an about dialog

   An About Dialog

A dialog that displays a simple message about the application, along with a
logo and copyright information.  This is designed to be suitable to be called
from an "About" menu item.

Color Dialog
============

.. figure:: images/color_dialog.png
   :scale: 50
   :alt: a color dialog on Mac OS

   A Color Dialog on Mac OS

A dialog that takes a |Color| object and displays a standard OS color-picker
dialog with an option that allows the setting of transparency/alpha if needed.

A convenience function |get_color| opens the dialog and returns a color value.

Confirmation Dialog
===================

.. figure:: images/confirmation_dialog.png
   :scale: 50
   :alt: a confirmation dialog

   A Confirmation Dialog

A dialog that presents a message and allows the user to select a Yes/No

> response, with the option to present a Cancel button as well.  The button
> which is the default, as well as the labels for the buttons can be controlled.

Reviewer: ```suggestion
response, with the option to present a Cancel button as well.  The default
button, as well as the labels for the buttons can be controlled.
```

Directory Dialog
================

A dialog that shows the standard OS file selection dialog for directories only,
as well as whether the dialog should follow "Open" or "Save As" conventions for
selecting a directory. Other options allow specifying a default path.

Typical usage for a save dialog looks like this::

    # display a directory dialog for opening a directory
    dialog = DirectoryDialog(parent=None)
    if dialog.open() == OK:
        print(f"Open {dialog.path}")

File Dialog
===========

A dialog that shows the standard OS file selection dialog, with options to
filter by file type as well as whether the dialog should follow "Open" or
"Save As" conventions.  Other options allow specifying a default directory and
file, and wildcards for filtering file type extensions.

Typical usage for a save dialog looks like this::

    # display a file dialog for saving a Python file
    dialog = FileDialog(parent=None, action="save as", wildcard="*.py")
    if dialog.open() == OK:
        print(f"Save to {dialog.path}")

Message Dialog
==============

.. figure:: images/error_dialog.png
   :scale: 50
   :alt: an error message dialog

   An Error Message Dialog

> A dialog that shows the a standard OS message dialog, where the user is only

Reviewer: ```suggestion
A dialog that shows the standard OS message dialog, where the user is only
```

There are convenience functions |information|, |warning| and |error| that
create a dialog of the appropriate severity level.

Progress Dialog
===============

.. figure:: images/progress_dialog.png
   :scale: 50
   :alt: a progress dialog

   A Progress Dialog

This class displays a progress bar in a dialog, together with a message and
various additonal optional information, such as time estimates.  The progress
dialog is designed to integrate with long-running computations that are using
the main thread, and make an effort to keep the UI at least minimally
responsive::

    def create_thumbnails(paths, size=(128, 128)):
        progress = ProgressDialog(
            title="Creating Thumbnails...",
            max=len(paths),
        )
        progress.open()

        for i, path in enumerate(paths):
            progress.message = f"Processing: {path}"
            file, ext = os.path.splitext(path)
            with Image.open(path) as im:
                im.thumbnail(size)
                im.save(file + "_thumbnail.png")
            (cont, skip) = progress.update(i)
            if not cont or skip:
                break

>         progress_update(len(paths))

Reviewer: ```suggestion
        progress.update(len(paths))
```

It can also be used with Python's standard library |Executor| classes to take
advantage of potential parallelism::

    def create_thumbnail(path, size):
        file, ext = os.path.splitext(path)
        with Image.open(path) as im:
            im.thumbnail(size)
            im.save(file + "_thumbnail.png")
        return path

    def create_thumbnails(paths, size=(128, 128)):
        executor = ThreadPoolExecutor()
        progress = ProgressDialog(
            title="Creating Thumbnails...",
            max=len(paths),
        )
        progress.open()

        for i, path in enumerate(map(create_thumbnail, paths)):
            progress.message = f"Processing: {path}"
            (cont, skip) = progress.update(i)
            if not cont or skip:
                break

        executor.shutdown(cancel_futures=True)

>         progress_update(len(paths))

Reviewer: ```suggestion
        progress.update(len(paths))
```

> This is contrasted with a situation where operations are handled using a

Reviewer: ```suggestion
This is in contrast with the situation where operations are handled using a
```
